### PR TITLE
[Builder] Loading source with local path fixes

### PR DESCRIPTION
**Pull request by alonmr**
1. relative path - ok for runtime, not for build time
2. abs path (/Users/alon/x.py, /v3io/users/alon/x.oy) - ok for runtime, for build time warn that it should be set as workdir and not as source (assumes it's already on the image and if download from v3io is needed the user should use v3io:// scheme).

### CODE REVIEW

**Review by alonmr — COMMENTED** [review]
on hold - waiting for https://github.com/mlrun/mlrun/pull/2867

File: mlrun/api/launcher.py [hunk 1]
```diff
@@ -146,9 +147,14 @@ def launch(
     def verify_base_image(runtime):
         pass
 
-    @staticmethod
-    def _enrich_runtime(runtime):
-        pass
+    def _enrich_runtime(self, runtime):
+        if runtime.metadata.project:
+            project = self.db.get_project(runtime.metadata.project)
```
Comment (liranbg) [inline mlrun/api/launcher.py:152]: do not use `self.db`, default to `runtime._get_db()` for now

**Review by liranbg — CHANGES_REQUESTED** [review]
minor

File: mlrun/api/launcher.py [hunk 2]
```diff
@@ -18,6 +18,7 @@
 import mlrun.common.schemas.schedule
 import mlrun.execution
 import mlrun.launcher.base
+import mlrun.projects.pipelines
```
Comment (Tankilevitch) [inline mlrun/api/launcher.py:21]: redundant

File: mlrun/api/utils/builder.py [hunk 3]
```diff
@@ -358,11 +358,10 @@ def build_image(
 
     context = "/context"
     to_mount = False
-    v3io = (
-        source.startswith("v3io://") or source.startswith("v3ios://")
-        if source
-        else None
-    )
+    v3io = False
```
Comment (Tankilevitch) [inline mlrun/api/utils/builder.py:361]: v3io is to broad, with boolean params I suggest using `is_` prefix e.g `is_v3io_source`

File: mlrun/projects/project.py [hunk 4]
```diff
@@ -1648,12 +1651,55 @@ def get_function(
...
-            function = enrich_function_object(
-                self, function, copy_function=copy_function
+            function = self.enrich_function_object(
+                function, copy_function=copy_function
             )
             self.spec._function_objects[key] = function
         return function
 
+    def enrich_function_object(
+        self, function, decorator=None, copy_function=True
```
Comment (Tankilevitch) [inline mlrun/projects/project.py:1661]: Few notes regarding the method:
- I don't think this method should be exposed to the user.
- Also I don't think this method needs to be part of the project object but rather a helper of some kind.
- Missing docstring

File: mlrun/api/launcher.py [hunk 5]
```diff
@@ -147,8 +148,14 @@ def verify_base_image(runtime):
...
-    def _enrich_runtime(runtime):
-        pass
+    def _enrich_runtime(runtime: "mlrun.runtimes.base.BaseRuntime"):
+        if runtime.metadata.project:
+            project = runtime._get_db().get_project(runtime.metadata.project)
+            # this is mainly for tests with nop db
+            # in normal use cases if no project is found we will get an error
+            if project:
+                project = mlrun.projects.project.MlrunProject.from_dict(project.dict())
+                project.enrich_function_object(runtime, copy_function=False)
```
Comment (Tankilevitch) [inline mlrun/api/launcher.py:158]: add comment why we enrich the function of object from the runtime only on server side

File: mlrun/api/launcher.py [hunk 6]
```diff
@@ -147,8 +148,14 @@ def verify_base_image(runtime):
         pass
 
     @staticmethod
-    def _enrich_runtime(runtime):
-        pass
+    def _enrich_runtime(runtime: "mlrun.runtimes.base.BaseRuntime"):
+        if runtime.metadata.project:
```
Comment (Tankilevitch) [inline mlrun/api/launcher.py:152]: in what cases runtime is not a part of a project in the API?

Comment (Tankilevitch) [reply mlrun/api/launcher.py:152]: runtimes are always executed as part of a project, you might need to set the project name prior to enriching the runtime or pass it if its not already been done.

File: mlrun/projects/project.py [hunk 7]
```diff
@@ -46,6 +46,7 @@
 from ..features import Feature
 from ..model import EntrypointParam, ModelObj
 from ..run import code_to_function, get_object, import_function, new_function
+from ..runtimes.pod import AutoMountType
```
Comment (Tankilevitch) [inline mlrun/projects/project.py:49]: -_-, we have multiple runtimes imports in this file, some are relative and some are from, please align

File: mlrun/run.py [hunk 8]
```diff
@@ -585,7 +585,7 @@ def new_function(
                      (job, mpijob, ..) the handler can also be specified in the `.run()` command, when not specified
                      the entire file will be executed (as main).
                      for nuclio functions the handler is in the form of module:function, defaults to "main:handler"
-    :param source:   valid path to git, zip, or tar file, e.g. `git://github.com/mlrun/something.git`,
+    :param source:   valid absolute path to git, zip, or tar file, e.g. `git://github.com/mlrun/something.git`,
                      `http://some/url/file.zip`
```
Comment (Tankilevitch) [inline mlrun/run.py:589]: I am not sure about the order of the comment, `absolute path to git` sounds a bit strange to me, wdyt?

File: mlrun/runtimes/kubejob.py [hunk 9]
```diff
@@ -62,18 +62,15 @@ def with_source_archive(
     ):
         """load the code from git/tar/zip archive at runtime or build
 
-        :param source:          valid path to git, zip, or tar file, e.g.
+        :param source:          valid absolute path or URL  to git, zip, or tar file, e.g.
```
Comment (Tankilevitch) [inline mlrun/runtimes/kubejob.py:65]: this comment is better, lets align it with all other docstring of source

File: mlrun/runtimes/local.py [hunk 10]
```diff
@@ -217,6 +217,8 @@ def _pre_run(self, runobj: RunObject, execution: MLClientCtx):
         execution._current_workdir = workdir
         execution._old_workdir = None
 
+        # _is_run_local is set when the user specifies local=True in run()
+        # in this case we don't want to extract the source code and contaminate the user's local dir
         if self.spec.build.source and not hasattr(self, "_is_run_local"):
```
Comment (Tankilevitch) [inline mlrun/runtimes/local.py:222]: 💯

File: mlrun/utils/helpers.py [hunk 11]
```diff
@@ -163,6 +163,32 @@ def verify_field_regex(
...
+    source: str, pull_at_runtime: bool = False, workdir: str = None
+):
+    if pull_at_runtime:
+        return
+
+    if "://" not in source:
+        if not path.isabs(source):
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                "Source must be a valid url or absolute path when 'pull_at_runtime' is False"
+            )
```
Comment (Tankilevitch) [inline mlrun/utils/helpers.py:176]: I would add suggestion on what are the alternative option

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Looks good, some fixes needed

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Looks good, some fixes needed

File: mlrun/api/launcher.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alonmr) [reply mlrun/api/launcher.py:158]: I guess the only IO that nop db does is logging a warning which we can also set verbose to False with tests and then we don't have any IO. We can add `kind` to its class (the interface db already has a kind attribute) but I don't like code where you check the kind of object you have so I prefer to keep it like this.

File: mlrun/api/launcher.py (hunk 6 above)
Comment (alonmr) [reply mlrun/api/launcher.py]: added project enrichment

File: mlrun/run.py [hunk 12]
```diff
@@ -585,8 +585,11 @@ def new_function(
                      (job, mpijob, ..) the handler can also be specified in the `.run()` command, when not specified
                      the entire file will be executed (as main).
                      for nuclio functions the handler is in the form of module:function, defaults to "main:handler"
-    :param source:   valid path to git, zip, or tar file, e.g. `git://github.com/mlrun/something.git`,
+    :param source:   valid absolute path or URL to git, zip, or tar file, e.g.
+                     `git://github.com/mlrun/something.git`,
```
Comment (Tankilevitch) [inline mlrun/run.py:589]:
```suggestion
    :param source:   valid absolute path to zip, tar file or URL to git, e.g.
                     `git://github.com/mlrun/something.git`,
```

**Review by Tankilevitch — APPROVED** [review]
Looks good to me, minor comment suggestion

**Pull request merged**